File: js/starfield.js
/* Nebula Strike — parallax starfield + drifting nebula background. */
(function (global) {
  "use strict";

  class Starfield {
    constructor(w, h) {
      this.w = w;
      this.h = h;
      this.layers = [];
      // Three parallax layers, far -> near.
      const config = [
        { count: 60, speed: 18, size: [0.6, 1.2], alpha: 0.5 },
        { count: 40, speed: 42, size: [1.0, 1.8], alpha: 0.7 },
        { count: 22, speed: 80, size: [1.6, 2.8], alpha: 1.0 },
      ];
      for (const c of config) {
        const stars = [];
        for (let i = 0; i < c.count; i++) {
          stars.push({
            x: Math.random() * w,
            y: Math.random() * h,
            r: Utils.rand(c.size[0], c.size[1]),
            tw: Utils.rand(0, Utils.TAU),
          });
        }
        this.layers.push({ stars, speed: c.speed, alpha: c.alpha });
      }
      // Nebula blobs for color depth.
      this.nebula = [];
      for (let i = 0; i < 4; i++) {
        this.nebula.push({
          x: Math.random() * w,
          y: Math.random() * h,
          r: Utils.rand(120, 260),
          hue: Utils.rand(180, 320),
          drift: Utils.rand(6, 16),
          phase: Utils.rand(0, Utils.TAU),
        });
      }
      this.t = 0;
    }

    update(dt) {
      this.t += dt;
      for (const layer of this.layers) {
        for (const s of layer.stars) {
          s.y += layer.speed * dt;
          s.tw += dt * 3;
          if (s.y > this.h + 4) {
            s.y = -4;
            s.x = Math.random() * this.w;
          }
        }
      }
      for (const n of this.nebula) {
        n.y += n.drift * dt;
        if (n.y - n.r > this.h) {
          n.y = -n.r;
          n.x = Math.random() * this.w;
          n.hue = Utils.rand(180, 320);
        }
      }
    }

    draw(ctx) {
      // Nebula glow.
      ctx.save();
      ctx.globalCompositeOperation = "lighter";
      for (const n of this.nebula) {
        const pulse = 0.5 + 0.5 * Math.sin(this.t * 0.5 + n.phase);
        const g = ctx.createRadialGradient(n.x, n.y, 0, n.x, n.y, n.r);
        g.addColorStop(0, Utils.hsl(n.hue, 70, 55, 0.12 + 0.05 * pulse));
        g.addColorStop(1, Utils.hsl(n.hue, 70, 30, 0));
        ctx.fillStyle = g;
        ctx.beginPath();
        ctx.arc(n.x, n.y, n.r, 0, Utils.TAU);
        ctx.fill();
      }
      ctx.restore();

      // Stars.
      for (const layer of this.layers) {
        for (const s of layer.stars) {
          const tw = 0.6 + 0.4 * Math.sin(s.tw);
          ctx.globalAlpha = layer.alpha * tw;
          ctx.fillStyle = "#cfe3ff";
          ctx.beginPath();
          ctx.arc(s.x, s.y, s.r, 0, Utils.TAU);
          ctx.fill();
        }
      }
      ctx.globalAlpha = 1;
    }
  }

  global.Starfield = Starfield;
})(window);

File: src/starfield.rs
use crate::utils::*;
use macroquad::prelude::*;

struct Star {
    pos: Vec2,
    speed: f32,
    size: f32,
    bright: f32,
}

/// Parallax scrolling star background drawn behind everything.
pub struct Starfield {
    stars: Vec<Star>,
    nebula_t: f32,
}

impl Starfield {
    pub fn new() -> Self {
        let mut stars = Vec::with_capacity(160);
        for _ in 0..160 {
            let depth = rand_range(0.2, 1.0);
            stars.push(Star {
                pos: vec2(rand_range(0.0, WIDTH), rand_range(0.0, HEIGHT)),
                speed: 20.0 + depth * 120.0,
                size: 0.5 + depth * 2.0,
                bright: 0.3 + depth * 0.7,
            });
        }
        Self {
            stars,
            nebula_t: 0.0,
        }
    }

    pub fn update(&mut self, dt: f32, scroll_boost: f32) {
        self.nebula_t += dt * 0.1;
        for s in &mut self.stars {
            s.pos.y += s.speed * (1.0 + scroll_boost) * dt;
            if s.pos.y > HEIGHT {
                s.pos.y = 0.0;
                s.pos.x = rand_range(0.0, WIDTH);
            }
        }
    }

    pub fn draw(&self) {
        // deep space gradient backdrop
        draw_rectangle(0.0, 0.0, WIDTH, HEIGHT, Color::new(0.02, 0.02, 0.06, 1.0));

        // soft drifting nebula blobs
        for i in 0..3 {
            let phase = self.nebula_t + i as f32 * 2.1;
            let x = WIDTH * (0.5 + 0.4 * (phase * 0.7).sin()) ;
            let y = HEIGHT * (0.5 + 0.4 * (phase * 0.5).cos());
            let col = match i {
                0 => Color::new(0.25, 0.05, 0.4, 0.05),
                1 => Color::new(0.05, 0.15, 0.4, 0.05),
                _ => Color::new(0.4, 0.1, 0.2, 0.04),
            };
            draw_circle(x, y, 260.0, col);
        }

        for s in &self.stars {
            draw_circle(s.pos.x, s.pos.y, s.size, with_alpha(WHITE, s.bright));
        }
    }
}
